Fix #4714: Alignment of button and text in exploration setting

saching97 opened the pull request:
## Explanation
Fixes #4714 
Implemented the solution of having the heading, button and text in separate lines as discussed in the issue thread.

## Screenshots
**Desktop view**
![alignment-fix-desktop](https://user-images.githubusercontent.com/24409931/37534427-d184d4ca-296a-11e8-97fd-21cab24e20e7.png)

**Mobile view**
![alignment-fix-mobile](https://user-images.githubusercontent.com/24409931/37534428-d221756e-296a-11e8-86cb-355aa6fffec4.png)

saching97 committed:
Fixes #4714

seanlip commented:
Thanks, @saching97! Hm ... tbh, I'm not sure that this is actually cleaner than the old version. Let's get some design input.

@christineycc: would you mind taking a look at #4714 and providing some feedback on whether this design is an improvement, or whether we should still keep the 3-column layout? Note that the editor page is only meant for desktop use.

saching97 commented:
Thanks @seanlip for the review. We can also choose to fix the width attribute of the button to 75px in the 3 columned layout and get aligned buttons and text as shown below. Maybe we can review both these options and go for the better one.
![alignment](https://user-images.githubusercontent.com/24409931/37599991-1a79bb26-2bac-11e8-8dca-35cf465cd543.png)

christineycc commented:
Hi,

Please see below feedback.

1. the buttons right now may be confusing and the status (enabled/disabled) is not very clear. Perhaps switch the buttons to something like below. And remove text that tries to indicate whether the function is enabled/disabled.

![frame 1](https://user-images.githubusercontent.com/25085761/37622621-89c7b0e6-2b98-11e8-87e6-949c6f488fd2.png)

2. The 3 column is squeezing the titles too much, while the button for the 1 column design is not as noticeable. For the layout, perhaps try something like the below, where title and button is on one line (Title aligned left and button aligned left), and description on the second line. And each option is separated by a light gray line.

Mobile:
![solution c mobile 1](https://user-images.githubusercontent.com/25085761/37624113-67e20abc-2b9d-11e8-8ed8-4a69d07a9124.png)
Desktop:
![solution b - 2 desktop 1](https://user-images.githubusercontent.com/25085761/37623903-ade96196-2b9c-11e8-8609-3f17b571196d.png)

If the above is hard to do, perhaps try putting the title on first line, the button on the second line, and description third line; and separate each title out by a gray line

I hope all of those are clear. Let me know if you have any questions!

saching97 committed:
Changed button and removed indicative text

saching97 commented:
Thanks for the review @christineycc. I have committed the mentioned changes, the screenshot is attached below. I've not modified the display for mobile view since @seanlip pointed out that the
> the editor page is only meant for desktop use.

![alignment-changed-icon](https://user-images.githubusercontent.com/24409931/37648553-49bce7b0-2c55-11e8-991d-84561f9ab310.png)

Also I noticed that the Parameters setting does not toggle (the function that is used there is `enableParameters()` unlike other functions which are `toggleAutomaticTextToSpeech()` and `toggleCorrectnessFeedback()`). That is once Parameter is enabled, it cannot be disabled. I'm opening a new issue for this.

saching97 committed:
Added protractor-test-enable-parameter class

saching97 committed:
Fixed toggle buttons

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -4876,6 +4876,72 @@ md-card.preview-conversation-skin-supplemental-card {
+  padding-right: 10px;
+  text-align: right;
+}
+.oppia-on-off-switch-switch-switch {
```
I'm not sure I understand the naming conventions here. What is switch-switch-switch?

Would it be possible to find clearer names?

seanlip commented:
Hi @saching97, I think this looks nice! @christineycc, do you have any final thoughts on it? (I just made one small inline code comment.)

Also, re the parameter thing -- that's actually kind of intended, I'm afraid. It's not straightforward to just disable the parameter functionality once it's enabled, especially if the user is using parameters in their explorations. So it would need to be a one-way switch for now.

christineycc commented:
I think it looks great! No other comments from me.

saching97 committed:
Fixed CSS names

saching97 commented:
I've updated the naming convention to be more understandable (@seanlip the previous was a result of an unrequited extra find and replace command). I've fixed the toggle buttons to show ON/OFF based on the value that they have and also made the Parameters as a one-way switch. 
Should I keep the issue #4823 open or close it for now?

seanlip reviewed (commented):
Looks good, thanks @saching97! I think you just need to pass the Travis checks.

And yep, we can close the other issue. I'll do that.

saching97 committed:
Removed extra protractor-test-enable-parameter class

saching97 commented:
Hi @seanlip I need a little help with the Travis build failing. I'm getting the following [error message](https://travis-ci.org/oppia/oppia/jobs/355874742#L2031)
```
Parameters
[chrome #01-1]     ✗ should navigate multiple states correctly, with parameters
[chrome #01-1]       - Failed: element not visible
```
...
`at ElementArrayFinder.applyAction_ (/home/travis/build/oppia/node_modules/protractor/lib/element.ts:482:23)
[chrome #01-1]         at ElementArrayFinder.(anonymous function) [as click] (/home/travis/build/oppia/node_modules/protractor/lib/element.ts:96:21)
[chrome #01-1]         at ElementFinder.(anonymous function) [as click] (/home/travis/build/oppia/node_modules/protractor/lib/element.ts:873:14)
[chrome #01-1]         at /home/travis/build/oppia/oppia/core/tests/protractor_utils/editor.js:860:59
[chrome #01-1]         at runFromSettingsTab (/home/travis/build/oppia/oppia/core/tests/protractor_utils/editor.js:803:16)
[chrome #01-1]         at Object.enableParameters (/home/travis/build/oppia/oppia/core/tests/protractor_utils/editor.js:859:3)
[chrome #01-1]         at UserContext.<anonymous> (/home/travis/build/oppia/oppia/core/tests/protractor/parameters.js:39:12)`

I concluded that this is due to the enableParameters not working as desired from editor.js. But I've checked and rechecked it and I'm still not able to understand the reason for the bug. The buttons are getting clicked (verified through Selenium) and the parameters do get enabled (I checked using the areParametersEnabled() function). If someone could help me out with this, it would be great. Thanks

seanlip commented:
Hi @saching97 -- thanks for your question! This is one of those things that needs some debugging, I think. The way I do this is to run the test manually on my local machine (see https://github.com/oppia/oppia/wiki/Running-Tests#end-to-end-tests for details) and actually look at what the browser is doing while following along with the code. (Note that you can run just the single test -- that will save you lots of time.) That helps me understand the exact location of the breakage.

Then, once I've found the location of the breakage, I tend to do something like `for (var i = 0; i < 100; i++) general.waitForSystem();` around there, so that the Protractor test runs up to that point, and waits. During that waiting period I poke around in the browser, inspect the console, etc., again to narrow down why the next action would fail.

Does this help? I'm sorry that I don't know the exact cause but the above steps are the sort of thing I would do to narrow down the problem if I was debugging this. I'd also use console logs etc. to help understand what is and isn't being called.

saching97 committed:
Moved protractor-test-enable-parameters

saching97 committed:
Fixed parameters error

saching97 commented:
Oh thanks a lot @seanlip! `for (var i = 0; i < 100; i++) general.waitForSystem();` was a lifesaver! The div with class "protractor-test-neutral-element" was not getting clicked during the test because of an overlapping problem. Changing it's size from 1x1 px to 3x3 px resolved all the issues :)

seanlip commented:
That's great @saching97, thanks for the explanation! The PR looks good -- merging now.

seanlip commented:
Also congrats on your first Oppia PR!
